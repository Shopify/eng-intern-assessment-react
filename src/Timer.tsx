import React, { Dispatch, SetStateAction, useEffect, useState } from 'react';
import { toStopWatchFormat } from './utils/time';
import { StopWatchAction, StopWatchState } from './resources/stopWatch';

interface IProps {
  state: StopWatchState;
  lapNumber: number;
  onLap: (time: number) => void;
}

export default function Timer({ state, lapNumber, onLap }: IProps) {
  const [time, setTime] = useState<number>(0);

  useEffect(() => {
    let intervalID: ReturnType<typeof setTimeout> | null;
    if (state === StopWatchState.RUNNING) {
      intervalID = setInterval(() => {
        setTime((prevTime) => prevTime + 1);
      }, 10);
    }
    if (state === StopWatchState.INITIAL) {
      setTime(0);
    }
    if (intervalID) return () => clearInterval(intervalID);
  }, [state]);

  useEffect(() => {
    if (lapNumber !== 0) onLap(time);
  }, [lapNumber]);

  return <div>{toStopWatchFormat(time)}</div>;
}
