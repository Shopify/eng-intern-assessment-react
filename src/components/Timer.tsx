import React from "react";
import dateFormatter from "../utils/TimeFormatter";

interface TimerProps {
  time: number;
  lapTime: number;
}

export default function Timer({ time, lapTime }: TimerProps) {
  return (
    <div className="timer">
      <div className="main-timer">
        <h2 className="digits" data-testid="main-timer">
          {dateFormatter.formatTimerTime(time)}
        </h2>
      </div>
      <div className="lap-timer">
        <h2 className="digits" data-testid="lap-timer">
          {dateFormatter.formatTimerTime(lapTime)}
        </h2>
      </div>
    </div>
  );
}
