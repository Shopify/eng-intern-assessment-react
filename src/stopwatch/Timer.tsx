import React, { useEffect, useState } from 'react';
import { toStopWatchFormat } from '../utils/time';
import { StopWatchState } from './resources/stopWatch';

interface IProps {
  state: StopWatchState;
  lapNumber: number;
  onLap: (time: number) => void;
}

// Attempts to minimize content rerendered when timer is updated
// by rerendering only the timer portion

export default function Timer({ state, lapNumber, onLap }: IProps) {
  const [time, setTime] = useState<number>(0);

  useEffect(() => {
    let intervalID: ReturnType<typeof setTimeout> | null;
    if (state === StopWatchState.RUNNING) {
      intervalID = setInterval(() => {
        setTime((prevTime) => prevTime + 1);
      }, 10); // increments timer every 10ms = 1/100 seconds
    }
    if (state === StopWatchState.INITIAL) {
      setTime(0);
    }
    if (intervalID) return () => clearInterval(intervalID);
  }, [state]);

  // Triggers lap handler when lap button is pressed
  // Does so by observing increases in lap #
  useEffect(() => {
    if (lapNumber !== 0) onLap(time);
  }, [lapNumber]);

  const { minutes, seconds, hundredthSec } = toStopWatchFormat(time);

  return (
    <div className='timer-container' data-testid='timer'>
      <span className='timer-number'>{minutes}</span>
      <span>:</span>
      <span className='timer-number'>{seconds}</span>
      <span>:</span>
      <span className='timer-number'>{hundredthSec}</span>
    </div>
  );
}
