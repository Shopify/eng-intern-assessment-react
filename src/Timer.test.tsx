import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import Timer from './Timer';

describe('Timer tests', () => {
    it('render should match snapshot file', () => {
        const tree = renderer.create(<Timer timeInMs={2500}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should result in correct time display value', () => {
        render(<Timer timeInMs={3500}/>);
        expect(screen.getByText('00:00:03.50')).toBeInTheDocument();
    });
});