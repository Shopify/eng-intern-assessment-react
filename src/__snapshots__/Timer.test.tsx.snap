// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer tests render should match snapshot file 1`] = `
<div>
  00:00:02.50
</div>
`;
